### PR TITLE
Call functions instead of evaluating strings

### PR DESCRIPTION
**Pull request by live627**
- do not evaluate strings to execute code
- ends usage of sSelf
- autosuggest callbacks are slightly reworked
- removing some old compatibility code

### CODE REVIEW

**Comment by live627** [comment]
The autosuggest plugin seems to support comma-separated values,  where it can send a search to the server based on the last value.

Screenshot 2023-11-09 at 12-58-32 Modify Boards

Not sure if this hidden feature should be removed.

**Comment by jdarwood007** [comment]
This may be outside the scope of what you intend to to here, but how much along does this help us to being able to use defer on our javascript files?  It couldn't happen for 2.1, but I would love it if we could add defer, aligning with modern suggestions for javascript to be either async or defer.  We would async the main script, defer the execution until the body is essentially loaded and then wait for dom to report its ready.

File: Themes/default/scripts/suggest.js [hunk 1]
```diff
@@ -338,24 +282,40 @@ smc_AutoSuggest.prototype.removeLastSearchString = function ()
...
 	// If there's a callback then call it.
-	if ('oCallback' in this && 'onBeforeAddItem' in this.oCallback && typeof(this.oCallback.onBeforeAddItem) == 'string')
+	if (typeof this.oCallback.onBeforeAddItem == 'function')
 	{
 		// If it returns false the item must not be added.
-		if (!eval(this.oCallback.onBeforeAddItem + '(' + this.opt.sSelf + ', \'' + sItemId + '\');'))
+		if (!this.oCallback.onBeforeAddItem.call(this, sItemId))
+			return;
+	}
+	// Backward compatibility; to be removed in the future
```
Comment (live627) [inline Themes/default/scripts/suggest.js:294]: This line is too vague. It needs to specify a milestone.

```suggestion
	// Backward compatibility; to be removed in v4.0.0.
```

Opinions? I hereby suggest the next major version.

**Comment by live627** [comment]
Beyond the scope of this branch. I am initially sending in my backlog of JavaScript changes. Script deferral https://github.com/SimpleMachines/SMF/issues/6539 comes in the future because the current setup isn't ready.

**Pull request closed without merging**